src/librustc_codegen_ssa/back/write.rs

```diff
@@ -390,10 +390,26 @@ pub fn start_async_codegen<B: ExtraBackendMethods>(
 
     for output_type in sess.opts.output_types.keys() {
         match *output_type {
-            OutputType::Bitcode => { modules_config.emit_bc = true; }
-            OutputType::LlvmAssembly => { modules_config.emit_ir = true; }
+            OutputType::Bitcode => {
+                modules_config.emit_bc = true;
+                if sess.opts.debugging_opts.no_link {
+                    metadata_config.emit_bc = true;
```

Comment:
I may be missing something here, but I don't really see why metadata codegen involves bitcode, LLVM IR, etc. This should only control the output of the `rmeta` file read by rustc if I'm not mistaken.

Reply:
I believe `-Zno-link` should still compile everything, only leaving the linking to a next rustc invocation.

Reply:
This metadata_config isn't for the `rmeta` file used later for linking. It is to [generate a metadata module linking into a `Dylib` or `ProcMacro`](https://github.com/rust-lang/rust/blob/f284f8b4be3a899bf2ecc03e2a1589f486b62a9f/src/librustc_interface/passes.rs#L911-L969). Bitcode/LLVM IR/etc. won't be generated for any other crate types.